### PR TITLE
Cleanup arrays and annotate for nullability

### PR DESCRIPTION
**Pull request by ds5678**

### CODE REVIEW

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs [hunk 1]
```diff
@@ -18,6 +24,13 @@ private protected static bool ThrowImmutableLength()
     {
         throw new NotSupportedException("Arrays have immutable length");
     }
+
+    private protected void ThrowIfIndexOutOfRange(int index)
+    {
+        if (index < 0 || index >= Length)
```
Comment (js6pak) [inline Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs:30]: dotnet/runtime uses the `(uint)index >= (uint)Length` pattern a lot[^1], should we do so aswell?

[^1]: https://github.com/search?q=repo%3Adotnet%2Fruntime+%22%28uint%29index+%3E%3D+%28uint%29%22&type=code

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs [hunk 2]
```diff
@@ -10,6 +11,11 @@ protected Il2CppArrayBase(IntPtr pointer) : base(pointer)
     {
     }
 
+    /// <summary>
+    /// The pointer to the first element in the array.
+    /// </summary>
+    private protected IntPtr ArrayStartPointer => IntPtr.Add(Pointer, 4 * IntPtr.Size);
```
Comment (js6pak) [inline Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs:17]:
```suggestion
    private protected unsafe IntPtr ArrayStartPointer => IntPtr.Add(Pointer, sizeof(Il2CppObject) /* base */ + sizeof(void*) /* bounds */ + sizeof(nuint) /* max_length */);
```

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppReferenceArray.cs [hunk 3]
```diff
@@ -41,27 +42,19 @@ public Il2CppReferenceArray(T[] arr) : base(AllocateArray(arr.Length))
...
-        }
+        get => WrapElement(GetElementPointer(index));
+        set => StoreValue(GetElementPointer(index), value?.Pointer ?? IntPtr.Zero);
+    }
+
+    private IntPtr GetElementPointer(int index)
+    {
+        ThrowIfIndexOutOfRange(index);
+        var elementPointer = IntPtr.Add(ArrayStartPointer, index * ourElementTypeSize);
+        return elementPointer;
```
Comment (js6pak) [inline Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppReferenceArray.cs:53]:
```suggestion
        return IntPtr.Add(ArrayStartPointer, index * ourElementTypeSize);
```

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppStringArray.cs [hunk 4]
```diff
@@ -25,27 +26,18 @@ public Il2CppStringArray(string[] arr) : base(AllocateArray(arr.Length))
...
-        set
-        {
-            if (index < 0 || index >= Length)
-                throw new ArgumentOutOfRangeException(nameof(index),
-                    "Array index may not be negative or above length of the array");
-            var arrayStartPointer = IntPtr.Add(Pointer, 4 * IntPtr.Size);
-            var elementPointer = IntPtr.Add(arrayStartPointer, index * IntPtr.Size);
-            *(IntPtr*)elementPointer = IL2CPP.ManagedStringToIl2Cpp(value);
-        }
+        get => IL2CPP.Il2CppStringToManaged(*GetElementPointer(index)) ?? "";
```
Comment (js6pak) [inline Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppStringArray.cs:29]: This shouldn't return empty string on null

**Review by js6pak — CHANGES_REQUESTED** [review]

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs [hunk 5]
```diff
@@ -143,7 +157,7 @@ public IndexEnumerator(Il2CppArrayBase<T> array)
 
         public void Dispose()
         {
-            myArray = null;
+            myArray = null!;
```
[the remark this answers is not in the document]
Comment (ds5678) [reply Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs:160]: Not really in my opinion

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs (hunk 1 above)
Comment (ds5678) [reply Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs]: Implemented

File: Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ds5678) [reply Il2CppInterop.Runtime/InteropTypes/Arrays/Il2CppArrayBase.cs:161]: But I don't want to change it either

**Review by js6pak — APPROVED** [review]

**Pull request merged**